File: leetcode/stack/150. Evaluate Reverse Polish Notation/cpp/main.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        
        unordered_map<string, function<int (int, int)>> map = {
            { "+" , [] (int a, int b) { return a + b; } },
            { "-" , [] (int a, int b) { return a - b; } },
            { "*" , [] (int a, int b) { return a * b; } },
            { "/" , [] (int a, int b) { return a / b; } },
        };
        
        stack<int> st;
        
        for(string& t : tokens) {
            if (!map.count(t)) {
                st.push(stoi(t));
            } else {
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                
                st.push(map[t](b, a));
            }
        }
    
        return st.top();
    }
};

File: leetcode/stack/150. Evaluate Reverse Polish Notation/py/main.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
      st = []
      operations = {
        "+": lambda a, b: a + b,
        "-": lambda a, b: a - b,
        "/": lambda a, b: int(a / b),
        "*": lambda a, b: a * b
      }
        
      for operator in tokens:
        if operator in '+-*/':
          operation = operations[operator]
          a = st.pop()
          b = st.pop()
          st.append(operation(b, a))
        else:
          st.append(int(operator))
          
      return st.pop()
